### ISSUE TITLE
Hubot message no longer visible since 0.67.0

### ISSUE THREAD

**Issue by giray** [RocketChat/Rocket.Chat#11552]
### Description:
Since the upgrade to 0.67.0 hubot instances have gone silent.  All bots are online, they receive messages, process and send a response.  However, no message is displayed in any channels or DMs

### Steps to reproduce:

1. upgrade to 0.67.0
2. type in private channel where hubot lives `hubot help`

### Expected behavior:
hubot should respond with help 

### Actual behavior:
No response from bot

### Server Setup Information:

- Version of Rocket.Chat Server: 0.67.0
- hubot@2.19.0
- hubot-rocketchat@1.0.12
- Operating System: docker rocketchat/rocket.chat latest
- Deployment Method: Docker
- Number of Running Instances:  1
- DB Replicaset Oplog: 
- NodeJS Version: 
- MongoDB Version: mongo:3.0

### Additional context
Other incoming and outgoing integration for fine.  3 bots that have been working well for more then a year have all of a sudden stopped.  Further investigation shows this only is applicable to private channels.  Public channels do not have this issue

**Comment by hobbez1** [comment]
Our hubot has similair issue. Direct messages to the bot does not work, but when running "bot help" in a public group, the bot does respond.

From the logs, it seems as if the room lookup has issues:
Failed DM:
hubot_1  | [Fri Jul 27 2018 07:49:40 GMT+0000 (UTC)] INFO Message receive callback id T9gw2dgqafGQHQbzj ts Fri Jul 27 2018 07:49:39 GMT+0000 (UTC)
hubot_1  | [Fri Jul 27 2018 07:49:40 GMT+0000 (UTC)] INFO [Incoming] ****: bot help
hubot_1  | [Fri Jul 27 2018 07:49:40 GMT+0000 (UTC)] INFO Message sent to hubot brain.
hubot_1  | [Fri Jul 27 2018 07:49:40 GMT+0000 (UTC)] INFO Sending Message To Room: <redacted string of chars>
hubot_1  | [Fri Jul 27 2018 07:49:40 GMT+0000 (UTC)] INFO Looking up Room ID for: <redacted string of chars>

Successful group message:
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO Message received with ID id
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO Message receive callback id n6NitFhQgfovCkopY ts Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO [Incoming] ****: bot help
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO Looking up Room Name for: F8CJyynGWt2eAMduT
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO setting roomName: testbot
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO Message sent to hubot brain.
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO Sending Message To Room: testbot
hubot_1  | [Fri Jul 27 2018 07:51:18 GMT+0000 (UTC)] INFO Looking up Room ID for: testbot

**Comment by ramrami** [comment]
@hobbez1 are you talking about Internal Hubot ? if it's the case you should enable it for DM 

![internal hubot](https://user-images.githubusercontent.com/9748435/37411648-f8c99bd0-27ab-11e8-8e81-c295b56a8787.png)

**Comment by hobbez1** [comment]
@ramrami 
No its not internal hubot, it it external running, and yes, it is set for DM
RESPOND_TO_DM=true

**Comment by giray** [comment]
So I have been trying to see what happens with tcpdump ... and whenever the bot tries to respond it sends 
```
{"msg":"updated","methods":["84"]}
```
and gets back

```
{
	"msg": "result",
	"id": "84",
	"error": {
		"isClientSafe": true,
		"error": "error-not-allowed",
		"reason": "Not allowed",
		"details": {
			"method": "getRoomIdByNameOrId"
		},
		"message": "Not allowed [error-not-allowed]",
		"errorType": "Meteor.Error"
	}
}
````

**Comment by giray** [comment]
This seems indeed to be due to **Remove cache layer and internal calculated property room.usernames ([#10749](https://github.com/RocketChat/Rocket.Chat/pull/10749 ))**  so the solutions if for the hubot-rocketchat adapter not to use the cache layer anymore

See https://github.com/RocketChat/hubot-rocketchat/issues/295

**Comment by timkinnane** [comment]
@giray You're totally right about where the issue comes from, but the cache layer in Hubot is something entirely different to what's referred to in those changes. This isn't an issue to be resolved in hubot.

Wish I'd seen this earlier. Been chasing it back all the way through the SDK. Have a fix to submit now though anyway...

In summary:
- #10749 removed usernames from the room object, so `getRoomIdByNameOrId` couldn't check for the user in the room anymore.
- Instead of finding another solution, the logic was simply removed and would only work then for public channels.
- I'm submitting a PR that uses a different auth method, that re-instates the check for user in room.

Related issues:
- https://github.com/RocketChat/Rocket.Chat/issues/11131
- https://github.com/RocketChat/Rocket.Chat.js.SDK/issues/21
- https://github.com/RocketChat/hubot-rocketchat/issues/295

PR incoming.

# HELP!

It would really help to have someone submit a regression test to Rocket.Chat core, to protect these methods from changing in ways that wipe out a bunch of peoples bots.

It would also be great if someone can go over the other methods in that commit and any other methods that do implement custom permission checking logic. They should all be using `RocketChat.authz.canAccessRoom` to keep things DRY and prevent such instances of point failures.

@rodrigok Any comment?

**Comment by stevenmcastano** [comment]
So is this something we need to grab a new version of the hubot adaptor for? Or do we need to wait for this to be merged into the main RockChat repo so we can recompile the entire chat application?

**Comment by timkinnane** [comment]
@stevenmcastano Need to wait for the merge. Currently nothing can be done in the SDK, bot or adapter. `getRoomIdByNameOrId ` was changed in core and won't return ID for private channels.

**Comment by timkinnane** [comment]
@stevenmcastano I also noticed the method is marked as deprecated. So I don't know what's going on there.

It might be that eventually the adapters need a new way of doing it, e.g. through the Rest API using the SDK, but that's a pretty big change and would effect a lot of legacy Hubots.

**Comment by reetp** [comment]
This is very sad.

Just wasted a day of my life trying to find out what was going wrong. I thought it was just me being stupid.

I have had more than one conversation with Rocket devs where they have said 'don't use the internal bot, we're going to deprecate it, use the external one blah....' and now they break the external one without a by your leave, just as I was in the process of migrating across.

Seriously ticked off.

Fortunately I don't rely on it, but those who do will either be stuck on old releases, or face breakages.

**Comment by animator013** [comment]
Is there any progress ?

**Comment by rodrigok** [comment]
@animator013 the PR from Tim (https://github.com/RocketChat/Rocket.Chat/pull/11703) was merged a few ours ago, so you can test our develop version or wait the next release (08/20 for candidate and 08/27 for final)

**Comment by erikyuzwa** [comment]
thanks @rodrigok and @timkinnane!

**Comment by timkinnane** [comment]
@rodrigok - I've found a repeat of the same problem in [joinRoom.js](https://github.com/RocketChat/Rocket.Chat/blob/186f791194126f814481f317ea5f6d8c2081837b/packages/rocketchat-lib/server/methods/joinRoom.js#L24-L26).

It is throwing an error for **all** attempts to join a private group, even when the bot/user is already in the group, simply because the new logic only checks if the room is type `c` and throws if not.

This isn't critical in the same way as the patch I already made, because the bots only call `joinRoom` for private rooms on startup because they might not know the difference between types of rooms they're told to join and which room's they already in.

Usually the error is caught and the bot can still operate in the room, because it was already joined. However, if the error is not caught it will throw and break the bot, with `error-not-allowed` for attempting to join a room they were already in.

I can submit a PR for this method, but I'm worried what other cases might be lurking. Since you changed the logic initially, you might have a better idea what else we need to fix.
